code/datums/ai/making_your_ai.md

```diff
@@ -0,0 +1,247 @@
+
+```dm
+/datum/ai_planning_subtree/item_throw_attack
+
+/datum/ai_planning_subtree/item_throw_attack/SelectBehaviors(datum/ai_controller/controller, delta_time)
+	var/obj/item/item_pawn = controller.pawn
+
+	if(!controller.blackboard[BB_ITEM_TARGET] || !DT_PROB(ITEM_AGGRO_ATTACK_CHANCE, delta_time))
+		return //no target, or didn't aggro
+
+	controller.queue_behavior(controller.blackboard[BB_ITEM_MOVE_AND_ATTACK_TYPE], BB_ITEM_TARGET, BB_ITEM_THROW_ATTEMPT_COUNT)
+	return SUBTREE_RETURN_FINISH_PLANNING
+```
+
+This subtree takes a blackboard named `BB_ITEM_TARGET`, the target of the item set by other subtrees, and if that exists alongside a probability to aggro, the subtree queues the behavior to attack that mob.
+
+So, neat. When you have a target, queue an attack. This item attack subtree is pretty basic, but a more complicated one may queue different attacks depending on the target. How does this fit into the subtrees?
+
+Let's look where it's used, specifically in the subtrees variable:
+
+```dm
+/datum/ai_controller/haunted
+    planning_subtrees = list(
+        ///this applies aggro for picking up the item
+        /datum/ai_planning_subtree/item_ghost_resist,
+        ///this picks targets from the aggro list
+        /datum/ai_planning_subtree/item_target_from_aggro_list,
+        ///this uses the target to attack.
+        /datum/ai_planning_subtree/item_throw_attack,
+    )
+```
+
+As you can see the subtrees go top to bottom on their processing. `SUBTREE_RETURN_FINISH_PLANNING` will prematurely end the subtrees, so we can be sure the ai will focus on the behaviors planned so far in a "priority list" kind of way.
+
+Let's visualize this in a case where the subtrees should stop prematurely!
+
+```dm
+/datum/ai_controller/haunted
+    planning_subtrees = list(
+        ///someone is currently holding the item,
+        ///preventing it from attacking!
+        ///resist and end planning.
+        /datum/ai_planning_subtree/item_ghost_resist,
+        ///this does not fire this time around
+        /datum/ai_planning_subtree/item_target_from_aggro_list,
+        ///this does not fire this time around
+        /datum/ai_planning_subtree/item_throw_attack,
+    )
+```
+
+### Subtree Setup
+
+Subtrees also have procs for when the mob first starts using them and when they stop. You can use this to make subtrees "reactive" to events via signals, and this is where we set defaults for blackboards if necessary (we want lists to be empty, not null!)
+
+Example:
+
+```
+/datum/ai_planning_subtree/item_ghost_resist/SetupSubtree(datum/ai_controller/controller)
+	RegisterSignal(controller.pawn, COMSIG_ITEM_EQUIPPED, .proc/on_equip)
+	controller.blackboard[BB_LIKES_EQUIPPER] = FALSE
+	controller.blackboard[BB_ITEM_AGGRO_LIST] = list()
+
+/datum/ai_planning_subtree/item_ghost_resist/ForgetSubtree(datum/ai_controller/controller)
+	UnregisterSignal(controller.pawn, list(COMSIG_ITEM_EQUIPPED, COMSIG_ITEM_DROPPED))
+```
+
+### Behaviors for subtrees
```

Comment:
for this you should use an example that uses blackboard keys and cleans them up in finish_action, there's some specific behavior that might be hard to grasp otherwise.

Reply:
good suggestion, will implement